### ISSUE TITLE
training process often stop with no error and memory is not released

### ISSUE THREAD

**Issue by yuanyuan961** [MIC-DKFZ/nnUNet#316]
Hi Isensee,

Thanks for your excellent work. I just start to run the 2d UNet code. But I got a very strange problem. I use the command:

nnUNet_train 2d nnUNetTrainerV2 TaskXXX_MYTASK FOLD

the training process seems fine at the beginning. But the training suddenly stop after 50 epochs with no prompts and no errors. The memory is still occupied without releasing. But the GPU is not occupied anymore. I try to use -c to continue training with another GPU. But it then stop after 96 epoch with the same situation(no prompts and no error with memory being occupied). The following is the training log when the stop happens:

![image](https://user-images.githubusercontent.com/70528355/92502673-cffd6c00-f232-11ea-8a95-a7af9b833ee1.png)
![image](https://user-images.githubusercontent.com/70528355/92502746-e9061d00-f232-11ea-983c-d9e4e53164fa.png)
![image](https://user-images.githubusercontent.com/70528355/92502916-16eb6180-f233-11ea-85f4-ee45270f34ae.png)
The GPU memory is still occupied but the training is already stopped. There is no corresponding process when run command: nvidia-smi, ps -l.

It seems it is not caused by some stopping criterion as I can still continue to train. I wonder if the code contain some prompts when the stopping criterion occurs?

Another quick question : is there one command for training with all five fold? Or I have to train one by one? I try to set FOLD parameter as list[0,1,2,3,4] but it does not work. Thanks in advance for your answer !

Best,
Yuan

**Comment by FabianIsensee** [comment]
Hi,
this is really interesting. The training being hung up like that suggests that the main process is waiting for batches from the background workers but is not getting any. But if any of the background workers were to crash then the others would get notified and the whole training would crash.
However, the fact that you are using 2D training points to a different potential problem. If you are using an Ubuntu operating system then there is a RAM (not GPU memory) leak when training with mixed precision. I have opened an issue with pytorch but unfortunately nobody has worked on it yet: https://github.com/pytorch/pytorch/issues/40911. That is out of my control.
If your RAM became full during training and the OS terminates background workers because of that then the terminated worker has no chance of communicating there is a problem with the others and therefore this situation would occur (I think).
So the solution would be to train without mixed precision by using the `--fp32` flag. Note that this issue does not exist in 3D networks and you can train that with mixed precision no problem.
Best,
Fabian

**Comment by yuanyuan961** [comment]
@FabianIsensee Thanks for your reply! Now I have two quick questions:
1) is there one command for training with all five fold? Or I have to train one by one? I try to set FOLD parameter as list[0,1,2,3,4] but it does not work. 
2) Why does not the inference process calculate the Dice score? I only got the prediction file .nii.gz but no quantification metric print? I use the following command:

nnUNet_predict -i INPUT_FOLDER -o OUTPUT_FOLDER -t TASK_NAME_OR_ID -m CONFIGURATION --save_npz

Do I miss some important parameters for calculating the inference Dice score?

Best,
Yuan

**Comment by FabianIsensee** [comment]
Hi,

1) no there is not. You need to do that yourself. This is intentional because you maybe want to train the folds at the same time on different GPUs and servers (you don't have to, it's just something you can do)
2) How would the inference command know what the ground truth is? Inference is intended to run predictions. There is no reference for unknown cases. If you are predicting a test set then you need to evaluate that yourself using one of the many publicly available tools. You can also use https://github.com/MIC-DKFZ/nnUNet/blob/38be208c38e0175e423c2ff9dc9554531989e48d/nnunet/evaluation/evaluator.py#L446

Best,
Fabian
